Added backend layer : Http Client, Hangfire, Data Client

vanyakot opened the pull request.

vanyakot committed:
First commit

maksimboch commented on Log_system/Data/Model/LogData.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
```
Лишние using. Студия сама их создает при генерации класса.
Проверь warning-и, обычно она пишет что using можно удалить.

В других классах тоже посмотри, в каждом не буду писать одинаковый комментарий.

vanyakot replied:
Удалил

maksimboch commented on Log_system/Data/Receive/Handler.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Log_system.Data.Receive
+{
+    public class Handler
```
Нужно более осмысленное название. Handler - это чистая абстракция. Ты должен из названия понимать что делает твой класс

vanyakot replied:
Переименовал

maksimboch commented on Log_system/Data/Receive/Handler.cs:
```diff
@@ -0,0 +1,29 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
```
У меня ругается на System.Text.Json. У тебя точно с ним работает?

vanyakot replied:
Да , работает , нужно скачать доп пакет , чтобы работало , мне студия изначально предлагала использовать using Newtonsoft.Json , но при десериализации просит не string , а Newtonsoft.Json.JsonReader.

maksimboch commented on Log_system/Data/Receive/Handler.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public class Handler
+    {
+        private IHttpClient _client;
```
Все, что инжектируем и создаем в конструкторе принято помечать как readonly

vanyakot replied:
Изменил на readonly

maksimboch commented on Log_system/Log_system.csproj:
```diff
@@ -0,0 +1,18 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <Folder Include="Controllers\" />
```
Знаю, что скоро здесь появится контроллер, но в целом не стоит оставлять пустые namespace в проекте.

vanyakot replied:
Убрал до реализации контроллера

maksimboch commented on Log_system/Log_system.csproj:
```diff
@@ -0,0 +1,18 @@
+  <ItemGroup>
+    <PackageReference Include="Hangfire" Version="1.7.11" />
+    <PackageReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.AspNetCore.Razor.Design" Version="2.2.0" PrivateAssets="All" />
```
Выглядит как лишняя зависимость, созданная из шаблона проекта. Проверь нужна ли она и удали если нет.

vanyakot replied:
Убрал

maksimboch commented on Log_system/Properties/launchSettings.json:
```diff
@@ -0,0 +1,29 @@
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+      "applicationUrl": "https://localhost:5001;http://localhost:5000"
```
Почему у application 2 урла? Это ты добавил?

vanyakot replied:
Нет , сразу было , оставил один

maksimboch commented on Log_system/Properties/launchSettings.json:
```diff
@@ -0,0 +1,29 @@
+    "Log_system": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "launchUrl": "api/values",
```
По идее если убрать эту строчку то перестанет при запуске открывать страницу. Проверь.

vanyakot replied:
Убрал

maksimboch commented on Log_system/Services/HttpClient/HttpClient.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Log_system.Services
+{
+    public class HttpClient:IHttpClient
```
HttpClient : IHttpClient
У нас для читаемости принято отбивать интерфейсы и тд пробелами.

vanyakot replied:
Отбил

maksimboch commented on Log_system/Services/HttpClient/HttpClient.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            _source = config["DataUrl"];
+        }
+        public async Task SendGetRequest()
```
Тут я бы тоже отбил новой строкой.

vanyakot replied:
Отбил

maksimboch commented on Log_system/Services/HttpClient/HttpClient.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public HttpClient(IConfiguration config)
+        {
+            _source = config["DataUrl"];
```
Секцию лучше получаеть через Configuration.GetSection().
По именованию тоже лучше назвать configuration.

vanyakot replied:
Поправил

maksimboch commented on Log_system/Services/HttpClient/HttpClient.cs:
```diff
@@ -0,0 +1,36 @@
+        }
+        public async Task SendGetRequest()
+        {
+            WebRequest request = WebRequest.Create(_source);
```
Я бы переписал на получение HttpResponseMessage. У больше заточен под получение контента

vanyakot replied:
Переписал

maksimboch commented on Log_system/Data/Receive/Handler.cs:
```diff
@@ -0,0 +1,29 @@
+
+        public void ReceiveInfo()
+        {
+            _client.SendGetRequest();
```
Метод у тебя async. Но используешь ты его синхронно. Почитай про async\await

vanyakot replied:
Переделал

maksimboch commented on Log_system/Services/HttpClient/IHttpClient.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public interface IHttpClient
+    {
+        string _response { get; set; }
```
Это лишнее в интерфейсе, тем более с геттером.

vanyakot replied:
Убрал

maksimboch commented on Log_system/appsettings.json:
```diff
@@ -0,0 +1,10 @@
+  },
+  "AllowedHosts": "*",
+  "DataUrl": "http://localhost:58297/log-system-data/get-log-system-data",
+  "LogDB":  "Data Source=DESKTOP-11J4L5C;Initial Catalog=LogData;Integrated Security=True;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"
```
Для connectionStrings обычно используют отдельную секцию и получают через Configuration.GetConnectionString()
https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.configurationextensions.getconnectionstring?view=dotnet-plat-ext-3.1

vanyakot replied:
Поправил

maksimboch commented on Log_system/Startup.cs:
```diff
@@ -0,0 +1,53 @@
+
+            app.UseHangfireDashboard();
+
+            RecurringJob.AddOrUpdate<Handler>(h => h.ReceiveInfo(), Cron.Minutely);
```
Я бы вынес создание job у hangire в отдельный метод CreateHangfireJobs()

vanyakot replied:
Вынес

maksimboch commented on Log_system/Startup.cs:
```diff
@@ -0,0 +1,53 @@
+            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+            services.AddHangfire(x => x.UseSqlServerStorage(@"Data Source=DESKTOP-11J4L5C;Initial Catalog=Query;Integrated Security=True;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"));
+            services.AddHangfireServer();
+            services.AddTransient<IHttpClient, HttpClient>();
```
Добавление сервисов тоже предлагаю вытащить в отдельный метод, а лучше даже класс.
Представь, что у тебя не 1, а скажем 50 сервисов. Это нормально захламит Startup файл

vanyakot replied:
Исправлено

maksimboch commented on Log_system/Startup.cs:
```diff
@@ -0,0 +1,53 @@
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc().SetCompatibilityVersion(CompatibilityVersion.Version_2_2);
+            services.AddHangfire(x => x.UseSqlServerStorage(@"Data Source=DESKTOP-11J4L5C;Initial Catalog=Query;Integrated Security=True;Encrypt=False;TrustServerCertificate=False;ApplicationIntent=ReadWrite;MultiSubnetFailover=False"));
```
ConnectionString надо вынести в конфиг.

vanyakot replied:
Вынес

maksimboch commented on Log_system/Services/Repositories/LogRepository.cs:
```diff
@@ -0,0 +1,23 @@
+        private string connectionString;
+        public LogRepository(IConfiguration config)
+        {
+            connectionString = config["LogData"];
```
Аналогично, используй GetConnectionString когда будешь реализовывать репозиторий.

vanyakot replied:
Изменил

maksimboch commented on Log_system/Data/Receive/Handler.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Log_system.Data.Receive
```
Ты выделил два крупных namespace Data и Services. И это точно не Data :)

vanyakot replied:
Переместил в Services

vanyakot committed:
Исправление ошибок

1. Сделал секцию ConnectionString в appsettings.json.
2. Убрал лишние using'и.
3. Поменял имя объекта для конфигурации с config на configuration. Сделал получения данных из этого объекта через GetSection() и GetConnectionString().
4. Вынес строку подключения к бд Hangfire'a в appsettings.json

vanyakot committed:
Правки по код ревью

1. Сделал свойства readonly.
2. Убрал папку Controllers до момент реализации контроллера
3. Отрекдактировал launchSettings.json
4. Вынес создание job у hangire в отдельный метод CreateHangfireJobs()
5. Cделал отдельный метод для добавления своих сервисов
6. Переделал HttpClient.
7. Добавил новый сервис DataConverter , но по нему есть вопросы (нужно созвониться)

vanyakot committed:
Решил проблему с DataConverter'ом

Теперь все должно работать

maksimboch commented on Log_system/Services/DataConverter/DataConverter.cs:
```diff
@@ -0,0 +1,37 @@
+                }
+            }
+            response.Close();
+            return JsonSerializer.Deserialize<T>(json);
```
У нас принято отбивать return новой строкой. 
![изображение](https://user-images.githubusercontent.com/65163033/82289233-00cc4380-99ad-11ea-8f2d-2b0fd1c35b91.png)

vanyakot replied:
Отбил

maksimboch commented on Log_system/Services/DataConverter/DataConverter.cs:
```diff
@@ -0,0 +1,37 @@
+            _client = client;
+        }
+
+        public async Task<T> WebResponseToObj<T>()
```
Название метода обычно содержит в себе глагол, полноценно описывающий выполняемую им задачу.

vanyakot replied:
Переименовал

maksimboch commented on Log_system/Services/DataController/DataController.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public async Task ReceiveInfo()
+        {
+           data = await _converter.WebResponseToObj<LogData>();
```
Что-то с отступами.

vanyakot replied:
Поправил

maksimboch commented on Log_system/Services/DataController/DataController.cs:
```diff
@@ -0,0 +1,21 @@
+    public class DataController
+    {
+        private readonly IDataConverter _converter;
+        private LogData data;
```
Не понимаю зачем ты создаешь приватные переменные в области класса, а не метода?

vanyakot replied:
Переместил

maksimboch commented on Log_system/Services/DataConverter/DataConverter.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Log_system.Services
+{
+    public class DataConverter : IDataConverter
```
Я бы назвал этот класс DataClient, думаю можно это в звонке обсудить

vanyakot replied:
Переименовал

maksimboch commented on Log_system/Services/DataConverter/DataConverter.cs:
```diff
@@ -0,0 +1,37 @@
+
+        private WebResponse response;
+
+        private string json;
```
Опять непонятная переменная за методом

vanyakot replied:
Переместил

maksimboch commented on Log_system/Services/HttpClient/HttpClient.cs:
```diff
@@ -0,0 +1,25 @@
+        public async Task<WebResponse> GetHttpResponseMessage()
+        {
+            WebRequest request = WebRequest.Create(_source);
+            return await request.GetResponseAsync();
```
По форматированию не забывай смотреть перед пушем

vanyakot replied:
А где тут ошибка в форматировании ?

maksimboch commented on Log_system/Services/Repositories/LogRepository.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public class LogRepository : ILogRepository
+    {
+        private readonly string connectionString;
```
По именованию в классах. 
Приватные переменные _veryImportantVar. 
Константы VeryImportantVar.
Публичные поля VeryImportantVar.

vanyakot replied:
Переименовал

maksimboch commented on Log_system/Startup.cs:
```diff
@@ -0,0 +1,56 @@
+            RecurringJob.AddOrUpdate<DataController>(h => h.ReceiveInfo(), Cron.Minutely);
+        }
+
+        private void ExtensionsServices(IServiceCollection services)
```
ExtensionService не подходит, тут скорее ConfigureServices

vanyakot replied:
Так не получится же назвать , назвал RegisterServices

maksimboch commented on Log_system/appsettings.json:
```diff
@@ -0,0 +1,13 @@
+    }
+  },
+  "AllowedHosts": "*",
+  "DataUrl": "http://localhost:58297/log-system-data/get-log-system-data",
```
Эти данные в конфиге чисто твои. Хотелось бы чтобы они были в твоем персональном конфиге.

vanyakot committed:
Исправление ошибок по код ревью
